### PR TITLE
Epoll bug fixes

### PR DESCRIPTION
**Pull request by karya0**

### CODE REVIEW

**Comment by karya0** [comment]
@rohgarg: I think this PR should be merged in the 2.5 and 2.4 branches as well. What say?

**Comment by rohgarg** [comment]
@karya0: I agree. I'm not sure about the commit: "Restore epoll _after_ file plugin" though. Do we need that in the DMTCP 2.x? I thought the main reason we needed this fix was because the restore happens in the opposite order in 3.0.

**Comment by rohgarg** [comment]
This fixes issue #405.

File: src/plugin/ipc/event/eventconnection.cpp [hunk 1]
```diff
@@ -77,8 +77,13 @@ EpollConnection::postRestart()
...
   JTRACE("Recreating epoll connection") (_fds[0]) (id());
-  int tempFd = _real_epoll_create(_size);
-  JASSERT(tempFd >= 0);
+  int tempFd;
+  if (_size != 0) {
+    tempFd = _real_epoll_create(_size);
+  } else {
+    tempFd = _real_epoll_create1(_flags);
+  }
+  JASSERT(tempFd >= 0) (_size) (JASSERT_ERRNO);
```
Comment (rohgarg) [inline src/plugin/ipc/event/eventconnection.cpp]: Should you also print `_flags` here?

File: src/plugin/ipc/event/eventconnection.cpp [hunk 2]
```diff
@@ -100,12 +105,11 @@ void
 EpollConnection::onCTL(int op, int fd, struct epoll_event *event)
 {
   JASSERT(((op == EPOLL_CTL_MOD || op == EPOLL_CTL_ADD) && event != NULL) ||
-          op == EPOLL_CTL_DEL) (epollType())
+          op == EPOLL_CTL_DEL)
```
Comment (rohgarg) [inline src/plugin/ipc/event/eventconnection.cpp]: Should we also print the `op` argument here?

File: src/plugin/ipc/event/eventconnection.h [hunk 3]
```diff
@@ -74,21 +74,18 @@ class EpollConnection : public Connection
...
-      EPOLL_WAIT
+      EPOLL_REGULAR,
     };
 
-    EpollConnection(int size, int type = EPOLL_CREATE)
+    EpollConnection(int size = 0, int flags = 0)
       : Connection(EPOLL),
-      _type(type),
-      _size(size)
+      _type(EPOLL_REGULAR),
```
Comment (rohgarg) [inline src/plugin/ipc/event/eventconnection.h]: Are we using the `_type` field in any meaningful way now? If not, should we just remove it?

File: src/plugin/ipc/ipc.cpp [hunk 4]
```diff
@@ -215,10 +215,10 @@ EXTERNC void
 dmtcp_initialize_plugin()
 {
   dmtcp_register_plugin(sshPlugin);
+  dmtcp_register_plugin(eventPlugin);
```
Comment (rohgarg) [inline src/plugin/ipc/ipc.cpp]: Could you please add a comment explaining why it needs to be called here? Probably, just copy paste the description from issue #405.

File: test/Makefile.in [hunk 5]
```diff
@@ -51,7 +51,7 @@ endif
 
 DMTCP_INCLUDE=${top_srcdir}/include
 
-CFLAGS += -I${DMTCP_INCLUDE} -DDMTCP -fPIC -g
+CFLAGS += -I${DMTCP_INCLUDE} -DDMTCP -fPIC -g -std=gnu99
```
Comment (rohgarg) [inline test/Makefile.in]: Why do we need this now?

File: src/plugin/ipc/event/eventconnection.h (hunk 3 above)
Comment (karya0) [inline src/plugin/ipc/event/eventconnection.h]: That's a member of the parent class. I'll restructure `Connection` to improve readability.

File: src/plugin/ipc/ipc.cpp (hunk 4 above)
Comment (karya0) [inline src/plugin/ipc/ipc.cpp]: I think we can skip this reordering. This is not relevant. The Epoll and File plugins are independent of each other.

File: test/Makefile.in (hunk 5 above)
Comment (karya0) [inline test/Makefile.in]: Because we want to use some modern features like inline for loop initializer, etc. This should have been in there all along :).

**Comment by karya0** [comment]
@rohgarg: I have updated addressed most of the comments and updated the PR. Please take a look.

File: test/Makefile.in (hunk 5 above)
Comment (rohgarg) [inline test/Makefile.in]: I meant why do we need this in the context of the current PR?

**Pull request merged**